### ISSUE TITLE
Depthwise separable conv with floating point 16 and batch norm is too slow.

### ISSUE THREAD

**Issue by mohsenghafoorian** [tensorflow/tensorflow#27285]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): **Yes**
- OS Platform and Distribution: **16.04.5 LTS (Xenial Xerus)**
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: **n.a.**
- TensorFlow installed from (source or binary): **source**
- TensorFlow version (use command below): **1.12.0**
- Python version: **3.6.7**
- Bazel version (if compiling from source): **0.21.0**
- GCC/Compiler version (if compiling from source): **5.4.0**
- CUDA/cuDNN version: **10.0 / 7.3.0**
- GPU model and memory: **v100 16gb**

**Describe the current behavior**
The depthwise separable convolution, unlike the normal convolution when used with floating point 16 is slower than floating point 32, but more importantly when it is used with batch norm and an optmizer (sgd, adam, etc.), it becomes very much (~40 times) slower than floating point 32. Another unexpected behavior is getting "core dumped" with most of the filter sizes, that only happens in case of separable conv with fp16, but not with the other combinations. 
Following is what I am getting from the given code:
```
Experiment {'fp16': False, 'sep_conv': False, 'bn': True, 'kernel_size': 4} started
and took 1945 ms
Experiment {'fp16': True, 'sep_conv': False, 'bn': True, 'kernel_size': 4} started
and took 1425 ms
Experiment {'fp16': False, 'sep_conv': True, 'bn': True, 'kernel_size': 4} started
and took 4430 ms
Experiment {'fp16': True, 'sep_conv': True, 'bn': True, 'kernel_size': 4} started
and took 161081 ms
Experiment {'fp16': True, 'sep_conv': True, 'bn': False, 'kernel_size': 4} started
and took 7191 ms
Experiment {'fp16': True, 'sep_conv': True, 'bn': False, 'kernel_size': 3} started
2019-03-29 13:20:30.522213: F tensorflow/stream_executor/stream_executor_pimpl.cc:712] Check failed: 0 == size % 4 (0 vs. 2)need 32-bit multiple size to fill with 32-bit pattern
Aborted (core dumped)
```

**Describe the expected behavior**
I expect the batch norm not to make it so much slower and not to get core dumped with different kernel sizes.
Also, I expected fp 16, to be faster than fp 32, (as is in normal conv).

**Code to reproduce the issue**
```
import time
import tensorflow as tf
from tensorflow.contrib import layers

depth = 50
width = 100
num_filters = 100
batch_size = 16
repeat = 3


def create_optimize_op(use_fp_16: bool, use_separable_conv: bool, use_batch_norm: bool, kernel_size: int,
                       exp_index: int) -> tf.Tensor:

    label = tf.random_normal(shape=[batch_size, width, width, num_filters],
                             dtype=tf.float16 if use_fp_16 else tf.float32)
    input_image = tf.random_normal(shape=[batch_size, width, width, 3], dtype=tf.float16 if use_fp_16 else tf.float32)
    head = input_image
    for layer_ind in range(depth):
        conv_f = layers.separable_conv2d if use_separable_conv else layers.conv2d
        head = conv_f(head, num_filters, (kernel_size, kernel_size), stride=(1, 1), padding="SAME", normalizer_fn=None,
                      reuse=tf.AUTO_REUSE, scope=f"{exp_index}layer{layer_ind}", data_format="NHWC")
        if use_batch_norm:
            head = tf.contrib.layers.batch_norm(head, scale=True, is_training=True)
    reshaped_label = tf.reshape(label, shape=[batch_size, -1])
    reshaped_pred = tf.reshape(head, shape=[batch_size, -1])
    loss = tf.losses.softmax_cross_entropy(reshaped_label, reshaped_pred)
    return tf.train.GradientDescentOptimizer(learning_rate=0.01).minimize(loss)


experiments = [{"fp16": False, "sep_conv": False, "bn": True, "kernel_size": 4},
               {"fp16": True, "sep_conv": False, "bn": True, "kernel_size": 4},
               {"fp16": False, "sep_conv": True, "bn": True, "kernel_size": 4},
               {"fp16": True, "sep_conv": True, "bn": True, "kernel_size": 4},
               {"fp16": True, "sep_conv": True, "bn": False, "kernel_size": 4},
               {"fp16": True, "sep_conv": True, "bn": False, "kernel_size": 3}]

with tf.Session() as session:
    for index, experiment in enumerate(experiments):
        step = create_optimize_op(experiment["fp16"], experiment["sep_conv"], experiment["bn"],
                                  experiment["kernel_size"], index)
        init_op = tf.global_variables_initializer()
        update_ops = tf.get_collection(tf.GraphKeys.UPDATE_OPS)
        print(f"Experiment {experiment} started")
        # a first warmup run.
        _ = session.run([init_op])
        _ = session.run([update_ops, step])

        start = time.clock()
        for i in range(repeat):
            _ = session.run([init_op])
            _ = session.run([update_ops, step])
        end = time.clock()
        time_taken = int((end - start) * 1000)
        print(f"and took {time_taken} ms")
```

**Comment by robieta** [comment]
I notice that you're using contrib layers. Can you test this with Keras layers and see if you see the same issue?

**Comment by mohsenghafoorian** [comment]
Thanks so much for your response. Changing to keras layers fixes the very slow (~40x slower) combination of depthwise separable conv, fp16, and normalization (see the following report). Still with depthwise separable conv, using fp16 is ~1.6 times slower than using fp32.
Also, the experiment gets 'core dumped' with different kernel sizes (3).

> Experiment {'fp16': False, 'sep_conv': False, 'bn': True, 'kernel_size': 4} started
and took 1021 ms
Experiment {'fp16': True, 'sep_conv': False, 'bn': True, 'kernel_size': 4} started
and took 464 ms
Experiment {'fp16': False, 'sep_conv': True, 'bn': True, 'kernel_size': 4} started
and took 1815 ms
Experiment {'fp16': True, 'sep_conv': True, 'bn': True, 'kernel_size': 4} started
and took 2957 ms
Experiment {'fp16': True, 'sep_conv': True, 'bn': False, 'kernel_size': 4} started
and took 2829 ms
Experiment {'fp16': True, 'sep_conv': True, 'bn': False, 'kernel_size': 3} started
2019-04-24 13:57:01.250352: F tensorflow/stream_executor/stream_executor_pimpl.cc:712] Check failed: 0 == size % 4 (0 vs. 2)need 32-bit multiple size to fill with 32-bit pattern
Aborted (core dumped)

**Comment by recrusader** [comment]
I have the same problem. FP16 separable_conv2D is super slow.

**Comment by qinyao-he** [comment]
Same problem. tf.nn.depthwise_conv2d in FP16 turns out to be several times slower than in FP32.

**Comment by x10000year** [comment]
any update?

**Comment by Meteorix** [comment]
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/depthwise_conv_op_gpu.h#L1766

It seems to be a bug here. In this case `num_filter_backprop = 3 * 3 * 3, sizeof(fp16) = 2`, so `ThenMemset32` will definitely raise error `0 == size % 4 (0 vs. 2)`. 

@jvishnuvardhan could you take a look at this?
